File: Servicos/rVendaProduto.cs
using System;
using Comum;
using Repositorios;
using Modelos;

namespace Servicos.nsVenda
{
    public class rVendaProduto : IsVendaProduto
    {
        private readonly IpVendaProduto _repo;
        public rVendaProduto(IpVendaProduto repo) { _repo = repo; }

        public ColecaoVendaProduto Listar()
        {
            try { return _repo.Listar(); }
            catch (ExcecaoNascomercio) { throw; }
            catch (Exception ex) { throw new ExcecaoNascomercio("Erro em Listar VendaProduto [" + ToString() + "] - " + ex.Message); }
        }

        public ColecaoVendaProduto Consultar(dVendaProduto dados)
        {
            try { return _repo.Consultar(dados); }
            catch (ExcecaoNascomercio) { throw; }
            catch (Exception ex) { throw new ExcecaoNascomercio("Erro em Consultar VendaProduto [" + ToString() + "] - " + ex.Message); }
        }

        public ColecaoVendaProduto ConsultarTroca(dVendaProduto dados)
        {
            try { return _repo.ConsultarTroca(dados); }
            catch (ExcecaoNascomercio) { throw; }
            catch (Exception ex) { throw new ExcecaoNascomercio("Erro em ConsultarTroca VendaProduto [" + ToString() + "] - " + ex.Message); }
        }

        public int Incluir(dVendaProduto dados)
        {
            try { return _repo.Incluir(dados); }
            catch (ExcecaoNascomercio) { throw; }
            catch (Exception ex) { throw new ExcecaoNascomercio("Erro em Incluir VendaProduto [" + ToString() + "] - " + ex.Message); }
        }

        public int Alterar(dVendaProduto dados)
        {
            try { return _repo.Alterar(dados); }
            catch (ExcecaoNascomercio) { throw; }
            catch (Exception ex) { throw new ExcecaoNascomercio("Erro em Alterar VendaProduto [" + ToString() + "] - " + ex.Message); }
        }

        public int Excluir(dVendaProduto dados)
        {
            try { return _repo.Excluir(dados); }
            catch (ExcecaoNascomercio) { throw; }
            catch (Exception ex) { throw new ExcecaoNascomercio("Erro em Excluir VendaProduto [" + ToString() + "] - " + ex.Message); }
        }

        public int ExcluirTroca(dVendaProduto dados)
        {
            try { return _repo.Excluir(dados); }
            catch (ExcecaoNascomercio) { throw; }
            catch (Exception ex) { throw new ExcecaoNascomercio("Erro em ExcluirTroca VendaProduto [" + ToString() + "] - " + ex.Message); }
        }

        public int ExcluirControle(int controle)
        {
            try { return _repo.ExcluirControle(controle); }
            catch (ExcecaoNascomercio) { throw; }
            catch (Exception ex) { throw new ExcecaoNascomercio("Erro em ExcluirControle VendaProduto [" + ToString() + "] - " + ex.Message); }
        }

        public int ExcluirControleTroca(int controle)
        {
            try { return _repo.ExcluirControleTroca(controle); }
            catch (ExcecaoNascomercio) { throw; }
            catch (Exception ex) { throw new ExcecaoNascomercio("Erro em ExcluirControleTroca VendaProduto [" + ToString() + "] - " + ex.Message); }
        }
    }
}

File: ncRegras/rVendaProduto.vb
Imports System.Transactions

Imports ncDados.nsVenda
Imports ncPersistencia.nsVenda
Imports ncComum.nsFuncoes
Imports ncComum.nsExcecao

Namespace nsVenda

    Public Class rVendaProduto

        '-- Métodos de controle ( Várias chamadas; Controle de transação )

        Public Function Listar() As ColecaoVendaProduto

            Dim retorno As ColecaoVendaProduto
            Dim persistencia As pVendaProduto
            Dim retornoPersistencia As ColecaoVendaProduto

            Try

                retorno = New ColecaoVendaProduto

                persistencia = New pVendaProduto
                retornoPersistencia = persistencia.Listar()

                If Not retornoPersistencia Is Nothing Then
                    If retornoPersistencia.Count > 0 Then
                        retorno.AddRange(retornoPersistencia)
                    Else
                        retorno = Nothing
                    End If
                Else
                    retorno = Nothing
                End If

            Catch ex As Exception

                retorno = Nothing
                Throw New ExcecaoNascomercio("Erro em Listar Venda [" & Me.ToString() & "] - " & ex.Message, ex)

            End Try

            Listar = retorno

        End Function

        Public Function Consultar(ByVal dados As dVendaProduto) As ColecaoVendaProduto

            Dim retorno As ColecaoVendaProduto
            Dim persistencia As pVendaProduto
            Dim retornoPersistencia As ColecaoVendaProduto

            Try

                retorno = New ColecaoVendaProduto

                persistencia = New pVendaProduto
                retornoPersistencia = persistencia.Consultar(dados)

                If Not retornoPersistencia Is Nothing Then
                    If retornoPersistencia.Count > 0 Then
                        retorno.AddRange(retornoPersistencia)
                    Else
                        retorno = Nothing
                    End If
                Else
                    retorno = Nothing
                End If

            Catch ex As Exception

                retorno = Nothing
                Throw New ExcecaoNascomercio("Erro em Consultar Venda [" & Me.ToString() & "] - " & ex.Message, ex)

            End Try

            Return (retorno)

        End Function

        Public Function ConsultarTroca(ByVal dados As dVendaProduto) As ColecaoVendaProduto

            Dim retorno As ColecaoVendaProduto
            Dim persistencia As pVendaProduto
            Dim retornoPersistencia As ColecaoVendaProduto

            Try

                retorno = New ColecaoVendaProduto

                persistencia = New pVendaProduto
                retornoPersistencia = persistencia.ConsultarTroca(dados)

                If Not retornoPersistencia Is Nothing Then
                    If retornoPersistencia.Count > 0 Then
                        retorno.AddRange(retornoPersistencia)
                    Else
                        retorno = Nothing
                    End If
                Else
                    retorno = Nothing
                End If

            Catch ex As Exception

                retorno = Nothing
                Throw New ExcecaoNascomercio("Erro em Consultar Venda [" & Me.ToString() & "] - " & ex.Message, ex)

            End Try

            Return retorno

        End Function

        Public Function Incluir(ByVal dados As dVendaProduto) As Integer

            Dim retorno As Integer
            Dim persistencia As pVendaProduto

            Try

                Using ts As New TransactionScope

                    persistencia = New pVendaProduto
                    retorno = persistencia.Incluir(dados)

                    ts.Complete()
                End Using

            Catch ex As Exception

                retorno = Nothing
                Throw New ExcecaoNascomercio("Erro em Incluir Venda [" & Me.ToString() & "] - " & ex.Message, ex)

            End Try

            Incluir = retorno

        End Function

        Public Function Alterar(ByVal dados As dVendaProduto) As Integer

            Dim retorno As Integer
            Dim persistencia As pVendaProduto

            Try

                Using ts As New TransactionScope
                    persistencia = New pVendaProduto
                    retorno = persistencia.Alterar(dados)

                    ts.Complete()
                End Using

            Catch ex As Exception

                retorno = Nothing
                Throw New ExcecaoNascomercio("Erro em Alterar Venda [" & Me.ToString() & "] - " & ex.Message, ex)

            End Try

            Alterar = retorno

        End Function

        Public Function Excluir(ByVal dados As dVendaProduto) As Integer

            Dim retorno As Integer
            Dim persistencia As pVendaProduto

            Try

                Using ts As New TransactionScope
                    persistencia = New pVendaProduto
                    retorno = persistencia.Excluir(dados)
                    ts.Complete()
                End Using

            Catch ex As Exception

                retorno = Nothing
                Throw New ExcecaoNascomercio("Erro em Excluir Venda [" & Me.ToString() & "] - " & ex.Message, ex)

            End Try

            Return retorno

        End Function

        Public Function ExcluirTroca(ByVal dados As dVendaProduto) As Integer

            Dim retorno As Integer
            Dim persistencia As pVendaProduto

            Try

                Using ts As New TransactionScope
                    persistencia = New pVendaProduto
                    retorno = persistencia.Excluir(dados)
                    ts.Complete()
                End Using

            Catch ex As Exception

                retorno = Nothing
                Throw New ExcecaoNascomercio("Erro em Excluir Troca [" & Me.ToString() & "] - " & ex.Message, ex)

            End Try

            Return retorno

        End Function

        Public Function ExcluirControle(ByVal controle As Integer) As Integer

            Dim retorno As Integer
            Dim persistencia As pVendaProduto

            Try

                Using ts As New TransactionScope
                    persistencia = New pVendaProduto
                    retorno = persistencia.ExcluirControle(controle)
                    ts.Complete()
                End Using

            Catch ex As Exception

                retorno = Nothing
                Throw New ExcecaoNascomercio("Erro em Excluir Venda [" & Me.ToString() & "] - " & ex.Message, ex)

            End Try

            Return retorno

        End Function

        Public Function ExcluirControleTroca(ByVal controle As Integer) As Integer

            Dim retorno As Integer
            Dim persistencia As pVendaProduto

            Try

                Using ts As New TransactionScope
                    persistencia = New pVendaProduto
                    retorno = persistencia.ExcluirControleTroca(controle)
                    ts.Complete()
                End Using

            Catch ex As Exception

                retorno = Nothing
                Throw New ExcecaoNascomercio("Erro em Excluir Troca [" & Me.ToString() & "] - " & ex.Message, ex)

            End Try

            Return retorno

        End Function

    End Class

End Namespace
